exchanges/kraken/kraken_websocket.go

```diff
@@ -777,63 +576,84 @@ func (k *Kraken) wsProcessTrades(channelData *WebsocketChannelData, data []inter
 	return trade.AddTradesToBuffer(k.Name, trades...)
 }
 
-// wsProcessOrderBook determines if the orderbook data is partial or update
-// Then sends to appropriate fun
-func (k *Kraken) wsProcessOrderBook(channelData *WebsocketChannelData, data map[string]interface{}) error {
+// wsProcessOrderBook handles both partial and full orderbook updates
+func (k *Kraken) wsProcessOrderBook(c string, response []any, pair currency.Pair) error {
+	key := &subscription.Subscription{
+		Channel: c,
+		Asset:   asset.Spot,
+		Pairs:   currency.Pairs{pair},
+	}
+	if err := fqChannelNameSub(key); err != nil {
+		return err
+	}
+	s := k.Websocket.GetSubscription(key)
+	if s == nil {
+		return fmt.Errorf("%w: %s %s %s", subscription.ErrNotFound, asset.Spot, c, pair)
+	}
+	if s.State() == subscription.UnsubscribingState {
+		// We only care if it's currently unsubscribing
+		return nil
+	}
+
+	ob, ok := response[1].(map[string]any)
+	if !ok {
+		return errors.New("received invalid orderbook data")
+	}
+
+	if len(response) == 5 {
+		ob2, ok2 := response[2].(map[string]any)
+		if !ok2 {
+			return errors.New("received invalid orderbook data")
+		}
+
+		// Squish both maps together to process
+		for k, v := range ob2 {
+			if _, ok := ob[k]; ok {
+				return errors.New("cannot merge maps, conflict is present")
+			}
+			ob[k] = v
+		}
+	}
 	// NOTE: Updates are a priority so check if it's an update first as we don't
 	// need multiple map lookups to check for snapshot.
-	askData, asksExist := data["a"].([]interface{})
-	bidData, bidsExist := data["b"].([]interface{})
+	askData, asksExist := ob["a"].([]interface{})
+	bidData, bidsExist := ob["b"].([]interface{})
 	if asksExist || bidsExist {
-		checksum, ok := data["c"].(string)
+		checksum, ok := ob["c"].(string)
 		if !ok {
 			return errors.New("could not process orderbook update checksum not found")
 		}
 
-		k.wsRequestMtx.Lock()
-		defer k.wsRequestMtx.Unlock()
-		err := k.wsProcessOrderBookUpdate(channelData, askData, bidData, checksum)
-		if err != nil {
-			outbound := channelData.Pair // Format required "XBT/USD"
-			outbound.Delimiter = "/"
-			go func(resub *subscription.Subscription) {
-				// This was locking the main websocket reader routine and a
-				// backlog occurred. So put this into it's own go routine.
-				errResub := k.Websocket.ResubscribeToChannel(resub)
-				if errResub != nil {
-					log.Errorf(log.WebsocketMgr,
-						"resubscription failure for %v: %v",
-						resub,
-						errResub)
+		err := k.wsProcessOrderBookUpdate(pair, askData, bidData, checksum)
+		if errors.Is(err, errInvalidChecksum) {
+			log.Debugf(log.Global, "Resubscribing to invalid %s orderbook", pair)
```

Comment:
```suggestion
			log.Debugf(log.Global, "%s Resubscribing to invalid %s orderbook", k.Name, pair)
```

Otherwise this is too ambiguous when multple exchanges are running

Reply:
Fixed 32d0ce7f3